src/NUnitFramework/tests/Constraints/PropertyTests.cs

```diff
@@ -206,4 +208,68 @@ public void FailureMessageContainsChainedConstraintMessage()
             Assert.IsTrue(textMessageWriter.ToString().Contains("Extra"));
         }
     }
+
+    [TestFixture(TestOf = typeof(PropertyConstraint))]
+    public class PropertyHierarchyTests
+    {
+        [Test]
+        public void PropertyExists_PropertyDefinedInDerivedClass_ShouldExist()
+        {
+            var sut = new PropertyConstraint(nameof(Derived.SomeProperty), new EqualConstraint(42));
+
+            var instance = (Base)new Derived();
+            var actual = sut.ApplyTo(instance);
+
+            Assert.That(actual, Has.Property(nameof(ConstraintResult.Status)).EqualTo(ConstraintStatus.Success));
+        }
+
+        class Base { }
+        class Derived : Base
+        {
+            public int SomeProperty { get; set; } = 42;
+        }
+
+
+        private int[] _array;
+        private PropertyConstraint _countPropertyConstraint;
+
+        [SetUp]
+        public void BeforeEveryTest()
+        {
+            _array = new[] { 1, 2, 3 };
+            _countPropertyConstraint = new PropertyConstraint(nameof(IList<object>.Count), new EqualConstraint(_array.Length));
```

Comment:
Can you do a `_countPropertyExistsConstraint` in each test showing that `PropertyExistsConstraint` also behaves the same?

`((IResolveConstraint)Has.Property("X").EqualTo(42)).Resolve()` returns PropertyConstraint, whereas `((IResolveConstraint)Has.Property("X")).Resolve()` returns PropertyExistsConstraint.

Reply:
ok, I can add additional preparation and assertion for each test that demonstrate property existence

Reply:
now we have same logic in 2 different classes (PropertyConstraint and PropertyExistsConstraint). My rule of thumb says "extract common code paths when it appears in 3 or more places" - so I'd let this slip unless we have similar logic in more places (i.e. places that could also benefit from new reasoning about properties)